### PR TITLE
Added persistent array implementation

### PR DESCRIPTION
**Pull request by dolphingarlic**
Also, should we maybe move this before introducing the persistent segtree?

### CODE REVIEW

**Comment by bqi343** [comment]
I was hoping for a path copying approach for persistent array before persistent segtree

**Comment by dolphingarlic** [comment]
Would that have the same implementation as a persistent segtree then? (Minus the range query)

**Comment by bqi343** [comment]
essentially yes

**Review by bqi343 — CHANGES_REQUESTED** [review]
^ above

**Comment by dolphingarlic** [comment]
I kept the fat nodes in because I think it's still worthwhile to know the alternative approaches, even if they are worse

File: content/6_Advanced/Persistent.mdx [hunk 1]
```diff
@@ -8,18 +8,141 @@ prerequisites:
...
+}
+
+int get_item(int index, int time) {
+	// Gets the array item at a given index and time
+	return query(roots[time], index);
+}
+
+void update_item(int index, int value, int prev_time, int curr_time) {
+	// Updates the array item at a given index and time
+	roots[curr_time] = new Node(roots[prev_time]); // First, copy the root
```
Comment (bqi343) [inline content/6_Advanced/Persistent.mdx:116]: ^ isn't this line not necessary?

File: content/6_Advanced/Persistent.mdx [hunk 2]
```diff
@@ -8,18 +8,141 @@ prerequisites:
...
+<CPPSection>
+
+```cpp
+struct Node {
+	int val;
+	Node *l, *r;
+
+	Node(ll x) : val(x), l(nullptr), r(nullptr) {}
+	Node(Node *ll, Node *rr) : val(0), l(ll), r(rr) {}
+	Node(Node *cp) : val(cp->val), l(cp->l), r(cp->r) {}
```
Comment (bqi343) [inline content/6_Advanced/Persistent.mdx:83]: same as below

File: content/6_Advanced/Persistent.mdx [hunk 3]
```diff
@@ -8,18 +8,141 @@ prerequisites:
...
+Path copying is **fully persistent**. 
+
+## Persistent Segment Tree
+
+Since persistent arrays with path copying are so similar to sparse segment
+trees, it's relatively straightforward to convert one into a persistent
+segment tree - just add range queries!
+
+However, note that it's not possible to support range updates with range queries
+in persistent segment trees, since lazy propagation doesn't play well with path
```
Comment (bqi343) [inline content/6_Advanced/Persistent.mdx:142]: actually, lazy is possible

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by bqi343 — APPROVED** [review]

**Pull request merged**